velox/functions/prestosql/ArrayFunctions.h

```diff
@@ -728,6 +728,38 @@ struct ArrayRemoveNullFunctionString {
   }
 };
 
+template <typename T>
+struct ArrayNGramFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T)
+
+  void call(
+      out_type<Array<Array<Generic<T1>>>>& out,
+      const arg_type<Array<Generic<T1>>>& input,
+      int64_t n) {
+    if (n <= 0) {
+      VELOX_USER_FAIL("N must be positive.");
+    }
+
+    if (n > input.size()) {
+      auto& newItem = out.add_item();
+      newItem.copy_from(input);
+      return;
+    }
+
+    for (auto i = 0; i <= input.size() - n; ++i) {
+      auto& newItem = out.add_item();
+      for (auto j = 0; j < n; ++j) {
+        if (input[i + j].has_value()) {
+          auto& newGranularItem = newItem.add_item();
+          newGranularItem.copy_from(input[i + j].value());
```

Comment:
I'm thinking that this function could benefit from zero-copy implementation (which would be possible if it is implemented as a vector function). Do you know how often this function is applied to arrays of complex types?

Reply:
So far I have seen usage mostly on string and integer arrays. That said, complex types can show up in workload. We can consider making it vector function if a lot of complex argument cases show up.